File: plugins/demonicGorilla/src/main/scala/com/fredplugins/demonicgorillaV2/DemonicGorilla.scala
package com.fredplugins.demonicgorillaV2

import ethanApiPlugin.EthanApiPlugin
import net.runelite.api.Actor
import net.runelite.api.Client
import net.runelite.api.HeadIcon
import net.runelite.api.NPC
import net.runelite.api.Player
import net.runelite.api.coords.LocalPoint
import net.runelite.api.coords.WorldArea
import net.runelite.api.coords.WorldPoint

import scala.util.chaining.*
import scala.util.Try
import scala.compiletime.uninitialized

//case class GorillaFrame(animation: Int, interacting: Player, worldArea: WorldArea, overheadIcon: HeadIcon)
//case class DemonicGorillaData(possibleAttackStyles: Seq[AttackStyle & RegularAttack],
//															nextAttackTick: Int,
//															recentProjectileId: Int,
//															attacksUntilSwitch: Int,
//
//															currentAnimation: Int,
//															lastAnimation: Int,
//															currentWorldArea:WorldArea,
//															lastWorldArea: WorldArea,
//															currentInteracting: Player,
//															lastInteracting: Player,
//
//														 ) {
//
//															 }
case class MemorizedNpcData(
	index:       Int,
	id:          Int,
	animationId: Int,
	interactingWith: Player,
	overheadIcon: HeadIcon,
	worldLoc:    WorldPoint,
	localPoint:  LocalPoint
) {
	def worldArea: WorldArea = new WorldArea(worldLoc, 1, 1)
}
case class MemorizedStateData(
	nextPossibleAttackStyles: Seq[AttackStyle & RegularAttack] = AttackStyle.RegularAttacks,
	nextAttackTick:           Int = -100,
	attacksUntilSwitch: Int = Attacks_per_Switch,
	recentProjectileId: Int = -1,
	initiatedCombat: Boolean          = false,
	takenDamageRecently          : Boolean = false,
	disabledMeleeMovementForTicks: Int     = 0,
	changedPrayerThisTick     : Boolean = false,
	changedAttackStyleThisTick: Boolean = false,
	changedAttackStyleLastTick: Boolean         = false,
	lastTickOverheadIcon      : HeadIcon | Null = null,
	lastTickAnimation         : Int             = -1,
	lastWorldArea  : WorldArea | Null = null,
	lastTickInteracting: Player | Null = null
)
class DemonicGorilla(npc: NPC)(using client: Client) {

	private var npcData: MemorizedNpcData = MemorizedNpcData(-1, -1, -1, Option(npc.getInteracting).collect {
		case n: Player => n
	}.orNull, null, null, null)

	private var stateData = MemorizedStateData()

	override def toString: String = {
		def printProduct(cc: Product, indent: Int): String = {
				cc.productElementNames.zip(cc.productIterator).map(a => s"  ${a._1} = ${a._2}")
					.mkString(s"${cc.productPrefix}(\n", ",\n", "\n)").linesIterator.toList.map(_.prependedAll(" ".repeat(indent))).mkString("\n").stripPrefix(" ".repeat(indent))
		}
		s"""DemonicGorilla($npc) {
			 |  npcData = ${printProduct(npcData, 12)}
			 |  stateData = ${printProduct(stateData, 12)}
			 |}
			 |""".stripMargin
	}
	//	override def toString: String = {
	//		s"""DemonicGorilla(${npc}) {
	//			 |  index = ${index}
	//			 |  nextPossibleAttackStyles = ${nextPossibleAttackStyles}
	//			 |  nextAttackTick = ${nextAttackTick}
	//			 |  attacksUntilSwitch = ${attacksUntilSwitch}
	//			 |  recentProjectileId = ${recentProjectileId}
	//			 |  takenDamageRecently = ${takenDamageRecently}
	//			 |  disabledMeleeMovementForTicks = ${disabledMeleeMovementForTicks}
	//			 |  changedPrayer = ${changedPrayerThisTick}
	//			 |  thisTick = {
	//			 |    changedAttackStyle = ${changedAttackStyleThisTick}
	//			 |    animation = ${getAnimation}
	//			 |    overhead = ${getOverheadIcon}
	//			 |    interacting = ${getInteracting}
	//			 |  }
	//			 |  lastTick = {
	//			 |    changedAttackStyle = ${changedAttackStyleLastTick}
	//			 |    animation = ${lastTickAnimation}
	//			 |    overhead = ${lastTickOverheadIcon}
	//			 |    interacting = ${lastTickInteracting}
	//			 |  }
	//			 |}
	//			 |""".stripMargin
	//	}
	//	private var nextPossibleAttackStyles: Seq[AttackStyle & RegularAttack] = AttackStyle.RegularAttacks
	//	private var nextAttackTick          : Int                              = -100
		def getNextAttackTick: Int = stateData.nextAttackTick
		def setNextAttackTick(value: Int): Unit = stateData = stateData.copy(nextAttackTick = value)

	//	def index: Int = npc.getIndex
	//	def localLocation: LocalPoint = npc.getLocalLocation

//	private var attacksUntilSwitch: Int = Attacks_per_Switch
//	private var recentProjectileId: Int = -1
//
//	private var takenDamageRecently          : Boolean = false
//	private var disabledMeleeMovementForTicks: Int     = 0
//
//	private var changedPrayerThisTick     : Boolean = false
//	private var changedAttackStyleThisTick: Boolean = false
//
//	private var changedAttackStyleLastTick: Boolean         = false
//	private var lastTickOverheadIcon      : HeadIcon | Null = null
//	private var lastTickAnimation         : Int             = 0



	def gameTick(): Unit = {
		val oldNpcData = npcData

		this.npcData = MemorizedNpcData(
			npc.getIndex,
			npc.getId,
			npc.getAnimation,
			Option(npc.getInteracting).collect {
				case n: Player => n
			}.orNull,
			EthanApiPlugin.getHeadIcon(npc),
			npc.getWorldLocation,
			npc.getLocalLocation
		)

		stateData = stateData.copy(
//			nextPossibleAttackStyles = ,
//			nextAttackTick = ,
//			attacksUntilSwitch = ,
//			recentProjectileId = ,
//			initiatedCombat = ,
//			takenDamageRecently = ,
//			disabledMeleeMovementForTicks = ,
//			changedPrayerThisTick = ,
//			changedAttackStyleThisTick = ,
//			changedAttackStyleLastTick = stateData.changedAttackStyleThisTick,
			lastTickAnimation = oldNpcData.animationId,
			lastTickOverheadIcon = oldNpcData.overheadIcon,
			lastTickInteracting = oldNpcData.interactingWith,
			lastWorldArea = oldNpcData.worldArea
		)

//		this.stateData = .lastTickAnimation = oldNpcData.animationId
		this.stateData.lastTickOverheadIcon
	}

	def getIndex: Int = npcData.index
	def getWorldArea: WorldArea = npcData.worldArea
	def getAnimationId: Int = npcData.animationId

	def getLastTickAnimation: Int = stateData.lastTickAnimation
	def setLastTickAnimation(value: Int): Unit = {
		stateData = stateData.copy(lastTickAnimation = value)
	}

	def getInteracting: Player | Null = {
		npcData.interactingWith
	}

	def isMatching(actor: Actor): Boolean = {
		actor match {
			case n: NPC => (n == this.npc)
			case _ => false
		}
	}

	def getWorldLocation: WorldPoint = npcData.worldLoc
	def getLocalLocation: LocalPoint = npcData.localPoint

	def isTakenDamageRecently: Boolean = stateData.takenDamageRecently
	def setTakenDamageRecently(v: Boolean): Unit = stateData = stateData.copy(takenDamageRecently = v)

	def getLastTickOverheadIcon: HeadIcon | Null = stateData.lastTickOverheadIcon
	def setLastTickOverheadIcon(icon: HeadIcon): Unit = stateData = stateData.copy(lastTickOverheadIcon = icon)

	def getOverheadIcon: HeadIcon | Null = npcData.overheadIcon
	def decrementAttacksUntilSwitch(): Unit = {
		setAttacksUntilSwitch(getAttacksUntilSwitch - 1)
	}
	def setAttacksUntilSwitch(v: Int): Unit = stateData = stateData.copy(attacksUntilSwitch = v)
	def getAttacksUntilSwitch: Int = stateData.attacksUntilSwitch

	def getNextPosibleAttackStyles: Seq[AttackStyle & RegularAttack] = stateData.nextPossibleAttackStyles

	def setNextPosibleAttackStyles(styles: Seq[AttackStyle & RegularAttack]): Unit = stateData = stateData.copy(
		nextPossibleAttackStyles = styles,
	)

	def filterPossibleAttackStyles(p: AttackStyle => Boolean): Unit = {
		val newAttackStyles = stateData.nextPossibleAttackStyles.filter(p)
		stateData = stateData.copy(
			nextPossibleAttackStyles = newAttackStyles
		)
	}

	def setRecentProjectileId(id: Int): Unit = stateData = stateData.copy(recentProjectileId = id)
	def getRecentProjectileId: Int = stateData.recentProjectileId
	def getLastTickInteracting: Player = stateData.lastTickInteracting
	def setLastTickInteracting(p: Player): Unit = stateData = stateData.copy(lastTickInteracting = p)
	def setInitiatedCombat(b: Boolean): Unit = stateData = stateData.copy(initiatedCombat = b)
	def isInitiatedCombat: Boolean = stateData.initiatedCombat
	def setChangedPrayerThisTick(b: Boolean): Unit = stateData = stateData.copy(changedPrayerThisTick = b)
	def isChangedPrayerThisTick: Boolean = stateData.changedPrayerThisTick


	def isChangedAttackStyleThisTick: Boolean = stateData.changedAttackStyleThisTick
	def setChangedAttackStyleThisTick(v: Boolean): Unit = stateData = stateData.copy(changedAttackStyleThisTick = v)
	def isChangedAttackStyleLastTick: Boolean = stateData.changedAttackStyleLastTick
	def setChangedAttackStyleLastTick(v: Boolean): Unit = stateData = stateData.copy(changedAttackStyleLastTick = v)
	def setLastWorldArea(wa: WorldArea): Unit = stateData = stateData.copy(lastWorldArea = wa)
	def getLastWorldArea: WorldArea = stateData.lastWorldArea
	def setDisabledMeleeMovementForTicks(i: Int): Unit = stateData = stateData.copy(disabledMeleeMovementForTicks = i)
	def getDisabledMeleeMovementForTicks: Int = stateData.disabledMeleeMovementForTicks
	def decrementDisabledMeleeMovementForTicks(): Unit = setDisabledMeleeMovementForTicks(getDisabledMeleeMovementForTicks-1)

}

File: plugins/demonicGorilla/src/main/java/com/theplug/kotori/demonicgorillas/DemonicGorilla.java
/*
 * Copyright (c) 2022, Kotori <https://github.com/OreoCupcakes/>
 * Copyright (c) 2018, Woox <https://github.com/wooxsolo>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package com.theplug.kotori.demonicgorillas;

import ethanApiPlugin.EthanApiPlugin;
import net.runelite.api.Actor;
import net.runelite.api.HeadIcon;
import net.runelite.api.NPC;
import net.runelite.api.coords.WorldArea;

import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

class DemonicGorilla {
	static final int MAX_ATTACK_RANGE = 10; // Needs <= 10 tiles to reach target
	static final int ATTACK_RATE = 5; // 5 ticks between each attack
	static final int ATTACKS_PER_SWITCH = 3; // 3 unsuccessful attacks per style switch

	static final int PROJECTILE_MAGIC_SPEED = 8; // Travels 8 tiles per tick
	static final int PROJECTILE_RANGED_SPEED = 6; // Travels 6 tiles per tick
	static final int PROJECTILE_MAGIC_DELAY = 12; // Requires an extra 12 tiles
	static final int PROJECTILE_RANGED_DELAY = 9; // Requires an extra 9 tiles

	static final AttackStyle[] ALL_REGULAR_ATTACK_STYLES =
		{
			AttackStyle.MELEE,
			AttackStyle.RANGED,
			AttackStyle.MAGIC
		};
	private final NPC npc;
	private List<AttackStyle> nextPossibleAttackStyles;

	private int attacksUntilSwitch;
	private int nextAttackTick;
	private boolean initiatedCombat;
	private boolean takenDamageRecently;
	private boolean changedAttackStyleThisTick;
	private boolean changedAttackStyleLastTick;
	private int disabledMeleeMovementForTicks;

	public String debugString() {
		return
		"Atk2Switch = " + String.valueOf(attacksUntilSwitch) + ", " +
		"nextAtkTick = " + String.valueOf(nextAttackTick) + ", " +
		"(initCbt, recDmg) = (" + String.valueOf(initiatedCombat) + ", " + String.valueOf(takenDamageRecently) + "), " +
		"ΔStyle(thisT, lastT) = (" + String.valueOf(changedAttackStyleThisTick) + ", " + String.valueOf(changedAttackStyleLastTick) + "), " +
		"DblMeleeMoveT = " +  String.valueOf(disabledMeleeMovementForTicks) + ", ";
	}

	private int lastTickAnimation;
	private WorldArea lastWorldArea;
	private Actor lastTickInteracting;
	private int recentProjectileId;
	private HeadIcon lastTickOverheadIcon;

	DemonicGorilla(NPC npc) {
		this.npc = npc;
		this.nextPossibleAttackStyles = List.of(AttackStyle.MELEE, AttackStyle.RANGED, AttackStyle.MAGIC);
		this.nextAttackTick = -100;
		this.attacksUntilSwitch = ATTACKS_PER_SWITCH;
		this.recentProjectileId = -1;
	}

	NPC getNpc() {
		return this.npc;
	}

	List<AttackStyle> getNextPossibleAttackStyles() {
		return this.nextPossibleAttackStyles;
	}

	boolean nextPossibleAttackStylesContains(Predicate<AttackStyle> predicate) {
		return this.nextPossibleAttackStyles.stream().anyMatch(predicate);
	}

	DemonicGorilla filterNextPossibleAttackStylesContains(Predicate<AttackStyle> predicate) {
		this.nextPossibleAttackStyles = this.nextPossibleAttackStyles.stream().filter(predicate).collect(Collectors.toUnmodifiableList());
		return this;
	}

	DemonicGorilla setNextPossibleAttackStyles(AttackStyle ... nextPossibleAttackStyles) {
		this.nextPossibleAttackStyles = Arrays.stream(nextPossibleAttackStyles).collect(Collectors.toUnmodifiableList());
		return this;
	}
	int getAttacksUntilSwitch() {
		return this.attacksUntilSwitch;
	}

	void setAttacksUntilSwitch(int attacksUntilSwitch) {
		this.attacksUntilSwitch = attacksUntilSwitch;
	}
	void resetAttacksUntilSwitch() {
		this.attacksUntilSwitch = DemonicGorilla.ATTACKS_PER_SWITCH;
	}
	void decrementAttacksUntilSwitch() {
		this.attacksUntilSwitch = attacksUntilSwitch - 1;
	}

	int getNextAttackTick() {
		return this.nextAttackTick;
	}

	void setNextAttackTick(int nextAttackTick) {
		this.nextAttackTick = nextAttackTick;
	}

	int getLastTickAnimation() {
		return this.lastTickAnimation;
	}

	void setLastTickAnimation(int lastTickAnimation) {
		this.lastTickAnimation = lastTickAnimation;
	}

	WorldArea getLastWorldArea() {
		return this.lastWorldArea;
	}

	void setLastWorldArea(WorldArea lastWorldArea) {
		this.lastWorldArea = lastWorldArea;
	}

	boolean isInitiatedCombat() {
		return this.initiatedCombat;
	}

	void setInitiatedCombat(boolean initiatedCombat) {
		this.initiatedCombat = initiatedCombat;
	}

	Actor getLastTickInteracting() {
		return this.lastTickInteracting;
	}

	void setLastTickInteracting(Actor lastTickInteracting) {
		this.lastTickInteracting = lastTickInteracting;
	}

	boolean isTakenDamageRecently() {
		return this.takenDamageRecently;
	}

	void setTakenDamageRecently(boolean takenDamageRecently) {
		this.takenDamageRecently = takenDamageRecently;
	}

	int getRecentProjectileId() {
		return this.recentProjectileId;
	}

	void setRecentProjectileId(int recentProjectileId) {
		this.recentProjectileId = recentProjectileId;
	}

//	boolean isChangedPrayerThisTick() {
//		return this.changedPrayerThisTick;
//	}
//
//	void setChangedPrayerThisTick(boolean changedPrayerThisTick) {
//		this.changedPrayerThisTick = changedPrayerThisTick;
//	}

	boolean isChangedAttackStyleThisTick() {
		return this.changedAttackStyleThisTick;
	}

	void setChangedAttackStyleThisTick(boolean changedAttackStyleThisTick) {
		this.changedAttackStyleThisTick = changedAttackStyleThisTick;
	}

	boolean isChangedAttackStyleLastTick() {
		return this.changedAttackStyleLastTick;
	}

	void setChangedAttackStyleLastTick(boolean changedAttackStyleLastTick) {
		this.changedAttackStyleLastTick = changedAttackStyleLastTick;
	}

	HeadIcon getLastTickOverheadIcon() {
		return this.lastTickOverheadIcon;
	}

	void setLastTickOverheadIcon(HeadIcon lastTickOverheadIcon) {
		this.lastTickOverheadIcon = lastTickOverheadIcon;
	}

	int getDisabledMeleeMovementForTicks() {
		return this.disabledMeleeMovementForTicks;
	}

	void setDisabledMeleeMovementForTicks(int disabledMeleeMovementForTicks) {
		this.disabledMeleeMovementForTicks = disabledMeleeMovementForTicks;
	}

	HeadIcon getOverheadIcon() {
		return EthanApiPlugin.getHeadIcon(npc);
	}

	void onGameTick() {
		if (isTakenDamageRecently()) {
			setInitiatedCombat(true);
		}

		if (getOverheadIcon() != getLastTickOverheadIcon()) {
			if (isChangedAttackStyleLastTick() ||
				isChangedAttackStyleThisTick()) {
				// Apparently if it changes attack style and changes
				// prayer on the same tick or 1 tick apart, it won't
				// be able to move for the next 2 ticks if it attempts
				// to melee
				setDisabledMeleeMovementForTicks(2);
			} else {
				// If it didn't change attack style lately,
				// it's only for the next 1 tick
				setDisabledMeleeMovementForTicks(1);
			}
		}
		setLastTickAnimation(getNpc().getAnimation());
		setLastWorldArea(getNpc().getWorldArea());
		setLastTickInteracting(getNpc().getInteracting());
		setTakenDamageRecently(false);
	//			gorilla.setChangedPrayerThisTick(false);
		setChangedAttackStyleLastTick(isChangedAttackStyleThisTick());
		setChangedAttackStyleThisTick(false);
		setLastTickOverheadIcon(getOverheadIcon());
		setRecentProjectileId(-1);
	}

	enum AttackStyle {
		MAGIC,
		RANGED,
		MELEE,
		BOULDER
	}
}